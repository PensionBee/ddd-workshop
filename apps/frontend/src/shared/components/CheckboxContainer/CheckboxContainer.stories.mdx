import { CheckboxContainer } from "./CheckboxContainer";
import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";

<Meta
  title="Shared/Checkbox Container/Docs"
  component={CheckboxContainer}
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Checkbox Container

## Props

<ArgsTable of={CheckboxContainer} />

## Usage & Customisation

### Default

<Canvas>
  <Story id="shared-checkbox-container--default" />
</Canvas>

### Round

<Canvas>
  <Story id="shared-checkbox-container--round" />
</Canvas>

### Disabled

<Canvas>
  <Story id="shared-checkbox-container--disabled" />
</Canvas>

### Error

<Canvas>
  <Story id="shared-checkbox-container--error" />
</Canvas>

### Default Active

<Canvas>
  <Story id="shared-checkbox-container--default-active" />
</Canvas>

### Round Active

<Canvas>
  <Story id="shared-checkbox-container--round-active" />
</Canvas>

### Disabled Active

<Canvas>
  <Story id="shared-checkbox-container--disabled-active" />
</Canvas>

### Error Active

<Canvas>
  <Story id="shared-checkbox-container--error-active" />
</Canvas>
