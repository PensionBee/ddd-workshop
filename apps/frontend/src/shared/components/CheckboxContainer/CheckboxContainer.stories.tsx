import type { ComponentMeta } from "@storybook/react";
import { useForm } from "react-hook-form";
import { CheckboxContainer } from "./CheckboxContainer";
import { Box } from "../Box/Box";

export default {
  title: "Shared/Checkbox Container",
  component: CheckboxContainer,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof CheckboxContainer>;

export const Default = () => {
  const { control } = useForm();

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control}>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const Round = () => {
  const { control } = useForm();

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control} variant="circle">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const Disabled = () => {
  const { control } = useForm();

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control} disabled>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const Error = () => {
  const { control } = useForm();

  return (
    <Box d="flex">
      <CheckboxContainer
        name="checkbox"
        control={control}
        error={"An error message"}
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const DefaultActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control}>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const RoundActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control} variant="circle">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const DisabledActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return (
    <Box d="flex">
      <CheckboxContainer name="checkbox" control={control} disabled>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};

export const ErrorActive = () => {
  const { control } = useForm({ defaultValues: { checkbox: true } });

  return (
    <Box d="flex">
      <CheckboxContainer
        name="checkbox"
        control={control}
        error={"An error message"}
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo laboriosam
        assumenda minima facere. Recusandae illum consequatur laboriosam quia
        inventore similique id rerum pariatur incidunt possimus eius modi
        repudiandae, temporibus excepturi.
      </CheckboxContainer>
    </Box>
  );
};
