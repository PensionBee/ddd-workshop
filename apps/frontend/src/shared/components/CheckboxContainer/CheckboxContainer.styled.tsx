import { theme } from "@shared/styling/theme";
import styled from "styled-components";

export const StyledCheckboxContainer = styled.div`
  display: flex;
  flex-wrap: nowrap;
  justify-content: flex-start;
  align-items: start;
  column-gap: 20px;
`;

export const StyledCheckboxContainerError = styled.div`
  background-color: ${() => theme.colors.red100};
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 5px;
  padding-bottom: 5px;
  margin-top: 15px;
  width: 100%;
`;
