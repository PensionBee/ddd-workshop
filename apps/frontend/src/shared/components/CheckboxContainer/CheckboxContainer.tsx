import {
  StyledCheckboxContainer,
  StyledCheckboxContainerError,
} from "./CheckboxContainer.styled";
import { Checkbox, type CheckboxProps } from "../Checkbox/Checkbox";
import { Box } from "../Box/Box";
import { BodySmallL, DisclaimerR } from "../Typography/Typography";

export type CheckboxContainerProps = CheckboxProps & {
  children: React.ReactNode;
};

export const CheckboxContainer: React.FC<CheckboxContainerProps> = ({
  children,
  error,
  ...props
}) => {
  return (
    <>
      <StyledCheckboxContainer data-rc="CheckboxContainer">
        <Box d="flex" mt={2}>
          <Checkbox error={error} {...props} />
        </Box>
        <BodySmallL>{children}</BodySmallL>
      </StyledCheckboxContainer>
      {error && (
        <StyledCheckboxContainerError>
          <DisclaimerR color="white">{error}</DisclaimerR>
        </StyledCheckboxContainerError>
      )}
    </>
  );
};
