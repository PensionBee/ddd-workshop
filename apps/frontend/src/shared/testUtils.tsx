import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
