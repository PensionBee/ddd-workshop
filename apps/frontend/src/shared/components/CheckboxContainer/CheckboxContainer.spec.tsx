import { render } from "@shared/testUtils";
import { vi } from "vitest";
import { useForm } from "react-hook-form";
import { CheckboxContainer } from "./CheckboxContainer";

const CHECKBOX_TEST_ID = "checkbox";
const CHECKBOX_NAME = "Checkbox name";
const CHECKBOX_CONTENT = "Checkbox content";

const CHECKBOX_ERROR = "An error message";

const CheckboxContainerForm = ({ hasError }: { hasError?: boolean }) => {
  const { control } = useForm();
  return (
    <CheckboxContainer
      data-testid={CHECKBOX_TEST_ID}
      name={CHECKBOX_NAME}
      control={control}
      error={hasError ? CHECKBOX_ERROR : undefined}
    >
      {CHECKBOX_CONTENT}
    </CheckboxContainer>
  );
};

describe("CheckboxContainer", () => {
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByRole, getByText, asFragment } = render(
        <CheckboxContainerForm />
      );
      const checkbox = getByRole("checkbox");

      expect(checkbox).toBeVisible();
      expect(checkbox).toHaveAttribute("data-state", "unchecked");

      expect(getByText(CHECKBOX_CONTENT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("Error", () => {
    it("should render correctly", () => {
      const { getByRole, getByText, asFragment } = render(
        <CheckboxContainerForm hasError />
      );
      const checkbox = getByRole("checkbox");

      expect(checkbox).toBeVisible();
      expect(checkbox).toHaveAttribute("data-state", "unchecked");

      expect(getByText(CHECKBOX_CONTENT)).toBeVisible();
      expect(getByText(CHECKBOX_ERROR)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
