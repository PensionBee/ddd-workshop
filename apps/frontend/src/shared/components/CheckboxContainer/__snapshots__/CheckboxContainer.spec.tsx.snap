// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckboxContainer > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft hpkFBl"
    data-rc="CheckboxContainer"
  >
    <div
      class="sc-gKXNOb dCWoxL"
    >
      <div
        class="sc-gKXNOb fMouIk"
        data-rc="Checkbox"
      >
        <button
          aria-checked="false"
          class="sc-hKMtDE lefQuz"
          data-state="unchecked"
          role="checkbox"
          type="button"
          value="on"
        />
      </div>
    </div>
    <span
      class="sc-iBkiHk fUcqyE"
      data-rc="BodySmallL"
    >
      Checkbox content
    </span>
  </div>
</DocumentFragment>
`;

exports[`CheckboxContainer > Error > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft hpkFBl"
    data-rc="CheckboxContainer"
  >
    <div
      class="sc-gKXNOb dCWoxL"
    >
      <div
        class="sc-gKXNOb fMouIk"
        data-rc="Checkbox"
      >
        <button
          aria-checked="false"
          class="sc-hKMtDE eqsvsm"
          data-state="unchecked"
          role="checkbox"
          type="button"
          value="on"
        />
      </div>
    </div>
    <span
      class="sc-iBkiHk fUcqyE"
      data-rc="BodySmallL"
    >
      Checkbox content
    </span>
  </div>
  <div
    class="sc-gsnScm kyJyUj"
  >
    <span
      class="sc-iBkiHk lgdqQg"
      data-rc="DisclaimerR"
    >
      An error message
    </span>
  </div>
</DocumentFragment>
`;

exports[`CheckboxContainer Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe OZwHh"
    data-rc="CheckboxContainer"
  >
    <div
      class="sc-gKPRtg chQDqR"
    >
      <div
        class="sc-gKPRtg bxbHsa"
        data-rc="Checkbox"
      >
        <button
          aria-checked="false"
          class="sc-hLBbgP foHXdo"
          data-state="unchecked"
          role="checkbox"
          type="button"
          value="on"
        />
      </div>
    </div>
    <span
      class="sc-iBYQkv dvOEjP"
      data-rc="BodySmallL"
    >
      Checkbox content
    </span>
  </div>
</DocumentFragment>
`;

exports[`CheckboxContainer Error should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe OZwHh"
    data-rc="CheckboxContainer"
  >
    <div
      class="sc-gKPRtg chQDqR"
    >
      <div
        class="sc-gKPRtg bxbHsa"
        data-rc="Checkbox"
      >
        <button
          aria-checked="false"
          class="sc-hLBbgP kaskJB"
          data-state="unchecked"
          role="checkbox"
          type="button"
          value="on"
        />
      </div>
    </div>
    <span
      class="sc-iBYQkv dvOEjP"
      data-rc="BodySmallL"
    >
      Checkbox content
    </span>
  </div>
  <div
    class="sc-gswNZR jKEHJh"
  >
    <span
      class="sc-iBYQkv kYUYYL"
      data-rc="DisclaimerR"
    >
      An error message
    </span>
  </div>
</DocumentFragment>
`;
